ipaserver/install/dsinstance.py

```diff
@@ -516,68 +517,76 @@ def __setup_sub_dict(self):
-
-        self.backup_state("serverid", self.serverid)
-        self.fstore.backup_file(paths.SYSCONFIG_DIRSRV)
-
-        self.sub_dict['BASEDC'] = self.realm.split('.')[0].lower()
-        base_txt = ipautil.template_str(BASE_TEMPLATE, self.sub_dict)
-        logger.debug("%s", base_txt)
-
-        target_fname = paths.DIRSRV_BOOT_LDIF
-        base_fd = open(target_fname, "w")
-        base_fd.write(base_txt)
-        base_fd.close()
-
-        # Must be readable for dirsrv
-        os.chmod(target_fname, 0o440)
-        os.chown(target_fname, pent.pw_uid, pent.pw_gid)
-
-        inf_txt = ipautil.template_str(INF_TEMPLATE, self.sub_dict)
-        logger.debug("writing inf template")
-        inf_fd = ipautil.write_tmp_file(inf_txt)
-        inf_txt = re.sub(r"RootDNPwd=.*\n", "", inf_txt)
-        logger.debug("%s", inf_txt)
-        args = [
-            paths.SETUP_DS_PL, "--silent",
-            "--logfile", "-",
-            "-f", inf_fd.name,
-        ]
-        logger.debug("calling setup-ds.pl")
-        try:
-            ipautil.run(args)
-            logger.debug("completed creating DS instance")
-        except ipautil.CalledProcessError as e:
-            raise RuntimeError("failed to create DS instance %s" % e)
-
-        # check for open port 389 from now on
-        self.open_ports.append(389)
-
-        inf_fd.close()
-        os.remove(paths.DIRSRV_BOOT_LDIF)
+        # The new installer is api driven. We can pass it a log function
+        # and it will use it. Because of this, we can pass verbose true,
+        # and allow our logger to control the display based on level.
+        sds = SetupDs(verbose=True, dryrun=False, log=logger)
+
+        # General environmental options.
+        general_options = General2Base(logger)
+        general_options.set('strict_host_checking', False)
+        # Check that our requested configuration is actually valid ...
+        general_options.verify()
+        general = general_options.collect()
+
+        # Slapd options, ie instance name.
+        slapd_options = Slapd2Base(logger)
+        slapd_options.set('instance_name', self.serverid)
+        slapd_options.set('root_password', self.dm_password)
+        slapd_options.verify()
+        slapd = slapd_options.collect()
+
+        # Create userroot. Note that the new install does NOT
+        # create sample entries, so this is *empty*.
+        userroot = {
+            'cn': 'userRoot',
+            'nsslapd-suffix': self.suffix.ldap_text()
+        }
+
+        backends = [userroot, ]
+
+        sds.create_from_args(general, slapd, backends, None)
+
+        # Now create the new domain root object in the format that IPA expects.
+        # Get the instance ....
+
+        inst = DirSrv(verbose=True, external_log=logger)
+        inst.remote_simple_allocate(
+            ldapuri=ipaldap.get_ldap_uri(self.fqdn),
+            password=self.dm_password)
+        # This actually opens the conn and binds.
+        inst.open()
```

Comment:
The instance keeps a connection open. We should clean and close the connection when we are done with it:

```
inst.open()
try:
    ipadomain = IpaDomain(inst, dn=self.suffix.ldap_text())
    ipadomain.create(properties={
        'dc': self.realm.split('.')[0].lower(),
        'info': 'IPA V2.0',
    })
finally:
    inst.close()
```

Reply:
Fixed.